src/Microsoft.ML.Tokenizers/Model/SentencePieceBaseModel.cs

```diff
@@ -0,0 +1,751 @@
+
+        public virtual string Decode(IEnumerable<int> ids, bool considerSpecialTokens)
+        {
+            if (ids is null)
+            {
+                throw new ArgumentNullException(nameof(ids));
+            }
+
+            using IEnumerator<int> enumerator = ids.GetEnumerator();
+            if (!enumerator.MoveNext())
+            {
+                return string.Empty;
+            }
+
+            ValueStringBuilder sb = new(stackalloc char[256]);
+
+            int bytesCount = -1;
+            byte[]? bytesPoolArray = null;
+            bool prefixRemoved = false;
+            int suffixIndex = -1;
+            char prefixSuffixChar = EscapeWhiteSpaces ? SentencePieceNormalizer.DummyPrefix : ' ';
+
+            if (enumerator.Current <= MaxByteId && ByteFallback)
+            {
+                // First token is a byte token.
+
+                while (enumerator.Current < ByteCodeToIdOffset)
+                {
+                    // It is possible listing some special tokens before the byte tokens in the tokenizer's data.
+                    TryDecodeAsSpecialToken(this, enumerator.Current, considerSpecialTokens, ref sb);
+
+                    // Skip control tokens.
+                    if (!enumerator.MoveNext())
+                    {
+                        return sb.ToString();
+                    }
+                }
+
+                if (enumerator.Current <= MaxByteId && ByteFallback)
+                {
+                    EncodeByte(enumerator.Current, OneByteUtf8EncodingMaxId, ByteCodeToIdOffset, ref bytesCount, ref bytesPoolArray, ref sb);
+                }
+                else if (!TryDecodeAsSpecialToken(this, enumerator.Current, considerSpecialTokens, ref sb) && TryMapIdToToken(enumerator.Current, out string? token))
+                {
+                    AppendTokenWithCheckingPrefix(AddDummyPrefix, TreatWhitespaceAsSuffix, token!, prefixSuffixChar, ref sb, ref prefixRemoved, ref suffixIndex);
+                }
+            }
+            else if (!TryDecodeAsSpecialToken(this, enumerator.Current, considerSpecialTokens, ref sb) && TryMapIdToToken(enumerator.Current, out string? token))
+            {
+                AppendTokenWithCheckingPrefix(AddDummyPrefix, TreatWhitespaceAsSuffix, token!, prefixSuffixChar, ref sb, ref prefixRemoved, ref suffixIndex);
+            }
+
+            char[]? charPoolArray = null;
+
+            while (enumerator.MoveNext())
+            {
+                if (enumerator.Current < ByteCodeToIdOffset)
+                {
+                    if (bytesCount >= 1)
+                    {
+                        FlushBytes(ref bytesCount, ref bytesPoolArray, ref charPoolArray, ref sb);
+                    }
+
+                    // It is possible listing some special tokens before the byte tokens in the tokenizer's data.
+                    TryDecodeAsSpecialToken(this, enumerator.Current, considerSpecialTokens, ref sb);
```

Comment:
We should try to avoid invoking Current twice. At a minimum it's an extra interface call, but a less-than-ideal interface implementation could also behave strangely.